contributors/design-proposals/auth/node-tls-bootstrap-via-tpm.md

```diff
@@ -0,0 +1,280 @@
+If existing certificate expires before rotation succeeds (e.g. when node is
+partitioned from cluster or suspended), the node requires operator
+intervention.  Rotation can be changed to use the TPM-based flow, eliminating
+the need for manual intervention.
+
+#### Incorporate PCR quoting
+
+PCR quotes can be added to the CSR. These quotes record the state of the
+machine (firmware, kernel parameters, etc). Master may validate these values
+during CSR approval.
+
+## Implementation options
+
+### Option 1: directly in Kubelet
+
+#### New package `tpm`
+
+Pckage `tpm` will have helper funcs for TPM-related functionality. Uses
+[github.com/google/go-tpm/tpm2](https://github.com/google/go-tpm/tree/master/tpm2)
+under the hood.
+
+```go
+package tpm
+
+type TPM struct {...}
+func Open(path) (*TPM, error)
+func (t *TPM) Close() error
+
+// AttestKey returns series of PEM blocks with attestation data.
+func (t *TPM) AttestKey(privateKey) (attestData, error)
+// VerifyCSR parses CSR PEM blocks with attestation data and verifies the
+// signature (used by Approver).
+func (t *TPM) VerifyCSR(csrData) error
+```
+
+#### New kubeconfig fields
+
+New fields for TPM interaction are added under `users` in `kubeconfig`
+([AuthInfo
+struct](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/api/types.go)):
+
+1. BootstrapCSRAttestation={none|tpm} - extra attestation mode for CSRs. none
+   is default. Field is behind a [feature
+   gate](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-field-in-existing-api-version).
+   Note: other values may be added in the future for alternative attestation
+   schemes.
+1. TPM (new type TPMConfig) - TPM-specific config fields
+1. TPM.Path string - path to TPM device (character device or unix socket)
+1. TPM.AIKCertIndex int - NVRAM index of AIK certificate
+1. TPM.AIKTemplateIndex int - NVRAM index of AIK template
+
+#### Kubelet code
+
+[bootstrap.LoadClientCert](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/certificate/bootstrap/bootstrap.go#L47)
+is the entry point.
+
+Split [csr.RequestNodeCertificate](https://github.com/kubernetes/client-go/blob/master/util/certificate/csr/csr.go#L51) into
+
+```go
+func MakeNodeCSR(privateKey, nodeName) (csrData, error)
+func RequestNodeCertificate(client, csrData, privateKey) (certData, error)
+```
+
+In `bootstap.LoadClientCert`:
+
+```
+csrData, err := csr.MakeNodeCSR(privateKey, nodeName)
+// handle err
+switch config.BootstrapCSRAttestation {
+    case "tpm":
+        attestData, err := tpm.AttestKey(privateKey)
+        // handle err
+        csrData = append(csrData, attestData...)
+}
+certData, err := csr.RequestNodeCertificate(client, csrData, privateKey)
+// handle err
+```
+
+Note: in this doc `privateKey` is always `crypto.PrivateKey`, not a PEM-encoded
+byte blob.
+
+### Option 2: exec authentication plugin
```

Comment:
This section isn't very fleshed out, but based on our conversations this requires the exec auth plugin being able to interact with opaque private keys stored on the TPM. Would it be easier and more general to add pkcs11 support to our clients for using private keys?

Reply:
In the case of exec plugin, TLS key would actually be created outside of the TPM and stored on disk. It would only be leaded in temporarily to do the signing (or "certification" in TPM lingo).
Keys created/stored in the TPM can't be exported so exec plugin wouldn't be able to pass it to kubelet.

Abstracting to PKCS11 is an interesting idea. I don't know if it would be easier from plugin author's perspective, but probably more universal. I'll think about this some more...

Reply:
@mikedanese had some thoughts on pkcs11.

Also, to clarify, this section isn't very fleshed out because exec plugin would do exactly what kubelet does currently plus changes described above. The code just happens to live in a different binary. Let me know if you'd like more detail here.

Reply:
pkcs11 is a huge spec covering way more then we need, requires dynamic loading of symbols from object files to use, and is just a painful API to work with IMO:

http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cs01/pkcs11-curr-v2.40-cs01.pdf

If I wanted to write a plugin that connected a key service to the kubelet TLS stack, I would much prefer a concise, designed to purpose API, vs implementing a pkcs11 backend.